File: lib/datos/media.dart
/// Descarga y resolución de las imágenes y GIFs de los ejercicios.
///
/// La media (miniaturas 180x180 y GIFs animados) es propiedad de **Gym visual**
/// y no se distribuye con este repositorio ni se empaqueta en el APK: se
/// descarga desde el dataset original a un directorio local, que está en el
/// .gitignore.
///
/// Condiciones de uso de la media, según el NOTICE del dataset:
///   - se distribuye únicamente a 180x180
///   - hay que mantener visible la atribución `© Gym visual — https://gymvisual.com/`
///
/// Ver https://gymvisual.com/content/3-terms-and-conditions-of-use
library;

import 'dart:async';
import 'dart:io';

import 'package:flutter/widgets.dart' show ImageProvider, FileImage, NetworkImage;
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';

import 'semilla.dart';

const baseRemota =
    'https://raw.githubusercontent.com/hasaneyldrm/exercises-dataset/main/';

/// Requisito de licencia: se pinta al pie de la ficha de ejercicio.
const atribucion = '© Gym visual — https://gymvisual.com/';

const _descargasSimultaneas = 16;
const _tiempoEspera = Duration(seconds: 30);

Directory? _directorio;

/// Localiza el directorio de media. Hay que llamarlo antes de usar [resolver].
///
/// En Android e iOS el directorio de trabajo no es escribible, así que la ruta
/// la decide path_provider.
Future<void> inicializarMedia() async {
  final soporte = await getApplicationSupportDirectory();
  _directorio = Directory('${soporte.path}/media');
  await _directorio!.create(recursive: true);
}

/// Fija el directorio a mano. Solo lo usan los tests.
void fijarDirectorioMedia(Directory? directorio) => _directorio = directorio;

/// URL en el repositorio original para una ruta como 'images/0001-2gPfomN.jpg'.
String urlRemota(String rutaRelativa) =>
    baseRemota + rutaRelativa.replaceFirst(RegExp('^/'), '');

/// Fichero en disco donde vive (o vivirá) esa media.
File? ficheroLocal(String rutaRelativa) {
  final base = _directorio;
  if (base == null) return null;
  return File('${base.path}/${rutaRelativa.replaceFirst(RegExp('^/'), '')}');
}

bool _estaDescargado(String rutaRelativa) {
  final fichero = ficheroLocal(rutaRelativa);
  if (fichero == null || !fichero.existsSync()) return false;
  return fichero.lengthSync() > 0;
}

/// Devuelve la imagen local si ya está descargada, o la remota si no.
///
/// Es lo que permite que la app funcione igual durante la descarga, si el
/// usuario la omite, o si falló a medias. Flutter anima los GIF en ambos casos.
ImageProvider? resolver(String? rutaRelativa) {
  if (rutaRelativa == null || rutaRelativa.isEmpty) return null;
  if (_estaDescargado(rutaRelativa)) {
    return FileImage(ficheroLocal(rutaRelativa)!);
  }
  return NetworkImage(urlRemota(rutaRelativa));
}

File? get _marcaCompleta {
  final base = _directorio;
  return base == null ? null : File('${base.path}/.complete');
}

/// True si ya se descargó todo el paquete de media.
bool get descargaCompleta => _marcaCompleta?.existsSync() ?? false;

Future<void> _marcarCompleta() async {
  final marca = _marcaCompleta;
  if (marca == null) return;
  await marca.parent.create(recursive: true);
  await marca.writeAsString('');
}

/// Rutas de media que aún no están en disco, dado el catálogo.
List<String> rutasPendientes(List<EjercicioJson> ejercicios) {
  final todas = <String>[];
  for (final ejercicio in ejercicios) {
    for (final ruta in [ejercicio.image, ejercicio.gif]) {
      if (ruta.isNotEmpty) todas.add(ruta);
    }
  }
  return [
    for (final ruta in todas)
      if (!_estaDescargado(ruta)) ruta,
  ];
}

/// Resultado de una tanda de descarga.
class ResultadoDescarga {
  const ResultadoDescarga(this.hechos, this.total, this.fallos);

  final int hechos;
  final int total;
  final int fallos;
}

/// Descarga un fichero si falta. Devuelve true si acabó estando disponible.
Future<bool> _descargarUno(String rutaRelativa, http.Client cliente) async {
  final destino = ficheroLocal(rutaRelativa);
  if (destino == null) return false;
  if (_estaDescargado(rutaRelativa)) return true;

  await destino.parent.create(recursive: true);
  // Se escribe a .part y se renombra al final para que un corte de red no deje
  // ficheros truncados que luego se den por buenos.
  final parcial = File('${destino.path}.part');
  try {
    final respuesta =
        await cliente.get(Uri.parse(urlRemota(rutaRelativa))).timeout(_tiempoEspera);
    if (respuesta.statusCode != 200) throw HttpException('${respuesta.statusCode}');
    await parcial.writeAsBytes(respuesta.bodyBytes);
    await parcial.rename(destino.path);
    return true;
  } on Exception {
    if (parcial.existsSync()) await parcial.delete();
    return false;
  }
}

/// Descarga toda la media que falte, en paralelo y de forma reanudable.
///
/// [alProgresar] se llama periódicamente con (hechos, total, fallos), y
/// [cancelado] permite abortar: los descargas en vuelo terminan y no se lanzan
/// más.
Future<ResultadoDescarga> descargarTodo(
  List<EjercicioJson> ejercicios, {
  void Function(int hechos, int total, int fallos)? alProgresar,
  bool Function()? cancelado,
}) async {
  final pendientes = rutasPendientes(ejercicios);
  final total = pendientes.length;
  if (total == 0) {
    await _marcarCompleta();
    alProgresar?.call(0, 0, 0);
    return const ResultadoDescarga(0, 0, 0);
  }

  var hechos = 0;
  var fallos = 0;
  var siguiente = 0;
  final cliente = http.Client();

  Future<void> trabajador() async {
    while (true) {
      if (cancelado?.call() ?? false) return;
      if (siguiente >= pendientes.length) return;
      final ruta = pendientes[siguiente++];

      final bien = await _descargarUno(ruta, cliente);
      hechos++;
      if (!bien) fallos++;
      if (hechos % 10 == 0 || hechos == total) {
        alProgresar?.call(hechos, total, fallos);
      }
    }
  }

  await Future.wait([
    for (var i = 0; i < _descargasSimultaneas; i++) trabajador(),
  ]);
  cliente.close();

  if (fallos == 0 && hechos == total) await _marcarCompleta();

  return ResultadoDescarga(hechos, total, fallos);
}

File: app/utils/media.py
"""Descarga y resolución de las imágenes y GIFs de los ejercicios.

La media (miniaturas 180x180 y GIFs animados) es propiedad de **Gym visual** y no
se distribuye con este repositorio: se descarga desde el dataset original al
directorio `media/`, que está en el .gitignore.

Condiciones de uso de la media, según el NOTICE del dataset:
  - se distribuye únicamente a 180x180
  - hay que mantener visible la atribución `© Gym visual — https://gymvisual.com/`
Ver https://gymvisual.com/content/3-terms-and-conditions-of-use
"""

import os
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests

from app.utils import almacenamiento

BASE_REMOTA = "https://raw.githubusercontent.com/hasaneyldrm/exercises-dataset/main/"
ATRIBUCION = "© Gym visual — https://gymvisual.com/"

DIRECTORIO = almacenamiento.directorio_media()
MARCA_COMPLETA = DIRECTORIO / ".complete"

_TIEMPO_ESPERA = 30
_HILOS = 16


def url_remota(ruta_relativa):
    """URL en el repositorio original para una ruta como 'images/0001-2gPfomN.jpg'."""
    return BASE_REMOTA + ruta_relativa.lstrip("/")


def ruta_local(ruta_relativa):
    """Ruta en disco donde vive (o vivirá) esa media."""
    return DIRECTORIO / ruta_relativa.lstrip("/")


def resolver(ruta_relativa):
    """Devuelve la ruta local si el fichero ya está descargado, o la URL remota.

    `ft.Image(src=...)` acepta ambas, así que la app funciona igual durante la
    descarga, si el usuario la omite, o si falló a medias.
    """
    if not ruta_relativa:
        return None
    local = ruta_local(ruta_relativa)
    if local.exists() and local.stat().st_size > 0:
        return str(local)
    return url_remota(ruta_relativa)


def descarga_completa():
    """True si ya se descargó todo el paquete de media."""
    return MARCA_COMPLETA.exists()


def _descargar_uno(ruta_relativa, sesion):
    """Descarga un fichero si falta. Devuelve True si el fichero está disponible."""
    destino = ruta_local(ruta_relativa)
    if destino.exists() and destino.stat().st_size > 0:
        return True

    destino.parent.mkdir(parents=True, exist_ok=True)
    # Se escribe a .part y se renombra al final para que un corte de red no deje
    # ficheros truncados que luego se den por buenos.
    parcial = destino.with_suffix(destino.suffix + ".part")
    try:
        respuesta = sesion.get(url_remota(ruta_relativa), timeout=_TIEMPO_ESPERA, stream=True)
        respuesta.raise_for_status()
        with open(parcial, "wb") as f:
            for trozo in respuesta.iter_content(chunk_size=65536):
                if trozo:
                    f.write(trozo)
        os.replace(parcial, destino)
        return True
    except Exception:
        parcial.unlink(missing_ok=True)
        return False


def rutas_pendientes(ejercicios):
    """Rutas de media que aún no están en disco, dada la lista del catálogo."""
    todas = []
    for ejercicio in ejercicios:
        for ruta in (ejercicio.get("image"), ejercicio.get("gif")):
            if ruta:
                todas.append(ruta)
    return [r for r in todas if not (ruta_local(r).exists() and ruta_local(r).stat().st_size > 0)]


def descargar_todo(ejercicios, al_progresar=None, cancelado=None):
    """Descarga toda la media que falte, en paralelo y de forma reanudable.

    :param ejercicios: lista de dicts del catálogo (con claves 'image' y 'gif')
    :param al_progresar: callback(hechos, total, fallos) llamado periódicamente
    :param cancelado: callable que devuelve True para abortar la descarga
    :return: (hechos, total, fallos)
    """
    pendientes = rutas_pendientes(ejercicios)
    total = len(pendientes)
    if total == 0:
        MARCA_COMPLETA.parent.mkdir(parents=True, exist_ok=True)
        MARCA_COMPLETA.touch()
        if al_progresar:
            al_progresar(0, 0, 0)
        return 0, 0, 0

    hechos = 0
    fallos = 0
    sesion = requests.Session()

    with ThreadPoolExecutor(max_workers=_HILOS) as pool:
        futuros = [pool.submit(_descargar_uno, ruta, sesion) for ruta in pendientes]
        for futuro in as_completed(futuros):
            hechos += 1
            if not futuro.result():
                fallos += 1
            if al_progresar and (hechos % 10 == 0 or hechos == total):
                al_progresar(hechos, total, fallos)
            if cancelado and cancelado():
                # cancel() solo funciona en los que aún no han arrancado, que es
                # justo lo que interesa: los 16 en vuelo terminan y se acabó.
                for pendiente in futuros:
                    pendiente.cancel()
                break

    sesion.close()

    if fallos == 0 and hechos == total:
        MARCA_COMPLETA.parent.mkdir(parents=True, exist_ok=True)
        MARCA_COMPLETA.touch()

    return hechos, total, fallos
